File: factory/src/adapters/artifact/postgres-artifact-store.ts
/**
 * Composed PostgreSQL {@link ArtifactStore} adapter.
 *
 * The store splits the two storage concerns of an artifact:
 *
 *   - the immutable binary payload lives in object storage (S3 / MinIO),
 *     content-addressed under `objects/<sha256-hex>`;
 *   - the authoritative metadata (statuses, retention window, legal hold and
 *     purge record) lives in the PostgreSQL `artifacts` row.
 *
 * `putArtifact` follows the *upload-then-commit* protocol (Jalon B Amendment 5):
 * the payload is uploaded to a staging key `uploads/<id>.part`, promoted to its
 * content-addressed key, verified in object storage, and only then is the
 * PostgreSQL row committed. A failed verification or commit therefore leaves no
 * authoritative row — only a reclaimable orphan in object storage.
 *
 * Every governance decision (`getArtifactMetadata`, `openArtifact`,
 * `deleteArtifact`, `purgeArtifact`, `setLegalHold`) reads the PostgreSQL row as
 * the source of truth and re-uses the shared pure governance helpers of the
 * in-memory adapter, so both implementations enforce identical rules.
 */

import type {
  ArtifactMetadata,
  ArtifactStore,
  OpenArtifactResult,
  PutArtifactParams,
} from '../../ports/artifact/artifact-store.js'
import { ARTIFACT_HASH_PREFIX, computeArtifactHash, createArtifactId } from './artifact-hash.js'
import {
  buildArtifactMetadata,
  isArtifactDestroyable,
  refreshArtifactMetadata,
  toArtifactBytes,
} from './memory-artifact-store.js'
import type { SqlClient } from '../persistence/sql/db.js'
import {
  createSqlArtifactMetadataRepository,
  type SqlArtifactMetadataRepository,
  type SqlArtifactMetadataRepositoryOptions,
} from '../persistence/sql/sql-artifact-metadata-repository.js'

const DEFAULT_UPLOAD_PREFIX = 'uploads'
const DEFAULT_OBJECT_PREFIX = 'objects'

/** Default retention window, in days, when none is requested nor configured. */
export const DEFAULT_ARTIFACT_RETENTION_DAYS = 90

/** Environment variable overriding the default retention window. */
export const ARTIFACT_RETENTION_DAYS_ENV = 'ARTIFACT_RETENTION_DAYS'

/**
 * Structural object-storage contract used by the store. The concrete
 * {@link S3ObjectClient} satisfies it, as does the offline fake used by tests.
 */
export interface ArtifactBlobClient {
  /** Stores an object at `key`. */
  putObject(key: string, body: Uint8Array, contentType?: string): Promise<void>
  /** Server-side copies `sourceKey` to `destinationKey`. */
  copyObject(sourceKey: string, destinationKey: string): Promise<void>
  /** Reads an object, or returns `null` when it does not exist. */
  getObject(key: string): Promise<{ stream: AsyncIterable<Uint8Array> } | null>
  /** Deletes an object. Returns `true` when a deletion happened. */
  deleteObject(key: string): Promise<boolean>
  /** Optional existence probe used to verify an object before commit. */
  headObject?(key: string): Promise<boolean>
  /** Optional listing used to reclaim orphaned staging uploads. */
  listObjectKeys?(prefix: string): Promise<string[]>
  /** Optional SigV4 presigner for readable payloads. */
  getSignedUrl?(key: string, options?: { expiresInSeconds?: number; now?: Date }): string
}

/** Configuration of {@link PostgresArtifactStore}. */
export interface PostgresArtifactStoreConfig {
  /** Object-storage client for the binary payloads. */
  client: ArtifactBlobClient
  /** Pre-built metadata repository; takes precedence over `sqlClient`. */
  repository?: SqlArtifactMetadataRepository
  /** SQL client used to build a metadata repository when none is provided. */
  sqlClient?: SqlClient
  /** Tenant / hierarchy scoping of the metadata repository. */
  organizationId?: string
  workstreamId?: string
  namespaceId?: string
  workflowId?: string
  /** Prefix of transient staging uploads (default `uploads`). */
  uploadPrefix?: string
  /** Prefix of content-addressed payloads (default `objects`). */
  objectPrefix?: string
  /** Explicit default retention window, overriding the environment. */
  retentionDays?: number
  /** Environment consulted for {@link ARTIFACT_RETENTION_DAYS_ENV}. */
  env?: NodeJS.ProcessEnv
  /** Injectable clock, primarily for deterministic tests. */
  now?: () => Date
}

/** Composed object-storage + PostgreSQL implementation of {@link ArtifactStore}. */
export class PostgresArtifactStore implements ArtifactStore {
  readonly #client: ArtifactBlobClient
  readonly #repository: SqlArtifactMetadataRepository
  readonly #uploadPrefix: string
  readonly #objectPrefix: string
  readonly #retentionDays: number | undefined
  readonly #env: NodeJS.ProcessEnv
  readonly #now: () => Date

  constructor(config: PostgresArtifactStoreConfig) {
    this.#client = config.client
    if (config.repository) {
      this.#repository = config.repository
    } else if (config.sqlClient) {
      const options: SqlArtifactMetadataRepositoryOptions = {
        ...(config.organizationId !== undefined ? { organizationId: config.organizationId } : {}),
        ...(config.workstreamId !== undefined ? { workstreamId: config.workstreamId } : {}),
        ...(config.namespaceId !== undefined ? { namespaceId: config.namespaceId } : {}),
        ...(config.workflowId !== undefined ? { workflowId: config.workflowId } : {}),
      }
      this.#repository = createSqlArtifactMetadataRepository(config.sqlClient, options)
    } else {
      throw new Error('PostgresArtifactStore requires a `repository` or a `sqlClient`')
    }
    this.#uploadPrefix = config.uploadPrefix ?? DEFAULT_UPLOAD_PREFIX
    this.#objectPrefix = config.objectPrefix ?? DEFAULT_OBJECT_PREFIX
    this.#retentionDays = config.retentionDays
    this.#env = config.env ?? process.env
    this.#now = config.now ?? (() => new Date())
  }

  async putArtifact(params: PutArtifactParams): Promise<ArtifactMetadata> {
    const now = this.#now()
    const data = toArtifactBytes(params.data)
    const id = createArtifactId()
    const hash = computeArtifactHash(data)
    const retentionDays = params.retentionDays ?? this.#defaultRetentionDays()
    const metadata = buildArtifactMetadata({
      id,
      owner: params.owner,
      contentType: params.contentType,
      data,
      retentionDays,
      now,
    })
    const stagingKey = this.#stagingKey(id)
    const contentKey = this.#contentKey(hash)

    // 1. Upload the payload to the staging key.
    await this.#client.putObject(stagingKey, Uint8Array.from(data), 'application/octet-stream')
    // 2. Promote the staging object to its immutable content-addressed key.
    await this.#client.copyObject(stagingKey, contentKey)
    // 3. Verify the object exists in object storage before the commit.
    if (!(await this.#objectExists(contentKey))) {
      throw new Error('ARTIFACT_OBJECT_VERIFICATION_FAILED')
    }
    // 4. Commit the authoritative metadata row in PostgreSQL.
    await this.#repository.saveMetadata(metadata, contentKey)
    // 5. Best-effort removal of the staging object.
    await this.#bestEffortDelete(stagingKey)
    return metadata
  }

  async getArtifactMetadata(artifactId: string): Promise<ArtifactMetadata | null> {
    const metadata = await this.#repository.getMetadata(artifactId)
    if (!metadata) return null
    return refreshArtifactMetadata(metadata, this.#now())
  }

  async openArtifact(artifactId: string): Promise<OpenArtifactResult | null> {
    const record = await this.#repository.getMetadataAndStorageKey(artifactId)
    if (!record) return null
    const metadata = refreshArtifactMetadata(record.metadata, this.#now())
    if (metadata.availabilityStatus === 'purged') return null
    const object = await this.#client.getObject(record.storageKey)
    if (!object) return null
    return { stream: object.stream, metadata }
  }

  async deleteArtifact(artifactId: string, reason?: string): Promise<boolean> {
    return this.#destroy(artifactId, reason ?? 'deleted')
  }

  /**
   * Returns a pre-signed URL granting temporary read access to an artifact's
   * payload, or `null` when the artifact is unknown, no longer readable
   * (`availabilityStatus !== 'available'`) or when the underlying blob client
   * cannot presign.
   *
   * PostgreSQL metadata stays authoritative: the storage key and availability
   * are read from the repository row, never inferred from the caller.
   */
  async getSignedUrl(artifactId: string, options?: { expiresInSeconds?: number; now?: Date }): Promise<string | null> {
    if (typeof this.#client.getSignedUrl !== 'function') return null
    const record = await this.#repository.getMetadataAndStorageKey(artifactId)
    if (!record) return null
    const metadata = refreshArtifactMetadata(record.metadata, options?.now ?? this.#now())
    if (metadata.availabilityStatus !== 'available') return null
    return this.#client.getSignedUrl(record.storageKey, options)
  }

  async purgeArtifact(artifactId: string, reason?: string): Promise<boolean> {
    return this.#destroy(artifactId, reason ?? 'retention-expired')
  }

  async setLegalHold(artifactId: string, legalHold: boolean, reason?: string): Promise<ArtifactMetadata | null> {
    const now = this.#now()
    return this.#repository.updateLegalHold(artifactId, legalHold, reason, now)
  }

  /**
   * Deletes staging objects left behind by interrupted uploads. Returns the
   * reclaimed keys. Mirrors {@link S3ArtifactStore.collectOrphanedUploads}.
   */
  async collectOrphanedUploads(): Promise<string[]> {
    if (typeof this.#client.listObjectKeys !== 'function') return []
    const keys = await this.#client.listObjectKeys(`${this.#uploadPrefix}/`)
    const reclaimed: string[] = []
    for (const key of keys) {
      if (await this.#bestEffortDelete(key)) reclaimed.push(key)
    }
    return reclaimed
  }

  async #destroy(artifactId: string, reason: string): Promise<boolean> {
    const record = await this.#repository.getMetadataAndStorageKey(artifactId)
    if (!record) return false
    const now = this.#now()
    const metadata = refreshArtifactMetadata(record.metadata, now)
    if (!isArtifactDestroyable(metadata, now)) return false
    const purged = await this.#repository.purgeArtifact(artifactId, reason, now)
    if (!purged) return false
    await this.#bestEffortDelete(record.storageKey)
    return true
  }

  #defaultRetentionDays(): number {
    if (this.#retentionDays !== undefined) return this.#retentionDays
    const parsed = Number.parseInt(this.#env[ARTIFACT_RETENTION_DAYS_ENV] ?? '', 10)
    return Number.isFinite(parsed) && parsed >= 0 ? parsed : DEFAULT_ARTIFACT_RETENTION_DAYS
  }

  async #objectExists(key: string): Promise<boolean> {
    if (typeof this.#client.headObject === 'function') return this.#client.headObject(key)
    const object = await this.#client.getObject(key)
    return object !== null
  }

  async #bestEffortDelete(key: string): Promise<boolean> {
    try {
      return await this.#client.deleteObject(key)
    } catch {
      return false
    }
  }

  #stagingKey(id: string): string {
    return `${this.#uploadPrefix}/${id}.part`
  }

  #contentKey(hash: string): string {
    const digest = hash.startsWith(`${ARTIFACT_HASH_PREFIX}:`) ? hash.slice(ARTIFACT_HASH_PREFIX.length + 1) : hash
    return `${this.#objectPrefix}/${digest}`
  }
}

/** Wires a composed object-storage + PostgreSQL artifact store. */
export function createPostgresArtifactStore(config: PostgresArtifactStoreConfig): PostgresArtifactStore {
  return new PostgresArtifactStore(config)
}

File: factory-service/src/main/kotlin/io/whozoss/factory/artifact/infrastructure/persistence/PostgresArtifactStore.kt
package io.whozoss.factory.artifact.infrastructure.persistence

import com.fasterxml.jackson.databind.ObjectMapper
import io.whozoss.factory.artifact.domain.ArtifactAvailabilityStatus
import io.whozoss.factory.artifact.domain.ArtifactGovernance
import io.whozoss.factory.artifact.domain.ArtifactHash
import io.whozoss.factory.artifact.domain.ArtifactMetadata
import io.whozoss.factory.artifact.domain.ArtifactRetentionStatus
import io.whozoss.factory.artifact.infrastructure.blob.ArtifactBlobClient
import io.whozoss.factory.artifact.port.ArtifactGcMetadataLister
import io.whozoss.factory.artifact.port.ArtifactGcMetadataRow
import io.whozoss.factory.artifact.port.ArtifactStore
import io.whozoss.factory.artifact.port.OpenArtifactResult
import io.whozoss.factory.artifact.port.PutArtifactParams
import io.whozoss.factory.persistence.TenantScope
import org.springframework.jdbc.core.JdbcTemplate
import org.springframework.jdbc.core.RowMapper
import java.sql.ResultSet
import java.sql.Timestamp
import java.time.Instant

/**
 * Composed PostgreSQL + object-storage implementation of [ArtifactStore].
 *
 * Ported from `factory/src/adapters/artifact/postgres-artifact-store.ts`. The
 * immutable binary payload lives in object storage, content-addressed under
 * `objects/<sha256-hex>`; the authoritative metadata (statuses, retention
 * window, legal hold and purge record) lives in the PostgreSQL `artifacts` row.
 *
 * `putArtifact` follows the *upload-then-commit* protocol: the payload is
 * uploaded to a staging key `uploads/<id>.part`, promoted to its
 * content-addressed key, verified in object storage, and only then is the
 * PostgreSQL row committed. A failed verification or commit therefore leaves no
 * authoritative row — only a reclaimable orphan in object storage.
 */
class PostgresArtifactStore(
    private val jdbcTemplate: JdbcTemplate,
    private val blobClient: ArtifactBlobClient,
    private val defaultRetentionDays: Int = ArtifactGovernance.DEFAULT_RETENTION_DAYS,
    private val uploadPrefix: String = DEFAULT_UPLOAD_PREFIX,
    private val objectPrefix: String = DEFAULT_OBJECT_PREFIX,
    private val objectMapper: ObjectMapper = ObjectMapper(),
    private val now: () -> Instant = Instant::now,
) : ArtifactStore, ArtifactGcMetadataLister {

    override fun putArtifact(scope: TenantScope, params: PutArtifactParams): ArtifactMetadata {
        val nowInstant = now()
        val id = ArtifactHash.createArtifactId()
        val hash = ArtifactHash.compute(params.data)
        val retentionDays = params.retentionDays ?: defaultRetentionDays
        val metadata = ArtifactGovernance.buildArtifactMetadata(
            id = id,
            owner = params.owner,
            contentType = params.contentType,
            data = params.data,
            retentionDays = retentionDays,
            now = nowInstant,
        )
        val stagingKey = "$uploadPrefix/$id.part"
        val contentKey = contentKey(hash)

        // 1. Upload the payload to the staging key.
        blobClient.putObject(stagingKey, params.data, "application/octet-stream")
        // 2. Promote the staging object to its immutable content-addressed key.
        blobClient.copyObject(stagingKey, contentKey)
        // 3. Verify the object exists in object storage before the commit.
        if (!blobClient.headObject(contentKey)) {
            throw IllegalStateException("ARTIFACT_OBJECT_VERIFICATION_FAILED")
        }
        // 4. Commit the authoritative metadata row in PostgreSQL.
        saveMetadata(scope, metadata, contentKey)
        // 5. Best-effort removal of the staging object.
        bestEffortDelete(stagingKey)
        return metadata
    }

    override fun getArtifactMetadata(scope: TenantScope, artifactId: String): ArtifactMetadata? {
        val row = selectRow(scope, artifactId) ?: return null
        return ArtifactGovernance.refreshArtifactMetadata(toMetadata(row), now())
    }

    override fun openArtifact(scope: TenantScope, artifactId: String): OpenArtifactResult? {
        val row = selectRow(scope, artifactId) ?: return null
        val metadata = ArtifactGovernance.refreshArtifactMetadata(toMetadata(row), now())
        if (metadata.availabilityStatus == ArtifactAvailabilityStatus.PURGED) return null
        val stream = blobClient.getObject(row.storageKey) ?: return null
        return OpenArtifactResult(stream = stream, metadata = metadata)
    }

    override fun deleteArtifact(scope: TenantScope, artifactId: String, reason: String?): Boolean =
        destroy(scope, artifactId, reason ?: "deleted")

    override fun purgeArtifact(scope: TenantScope, artifactId: String, reason: String?): Boolean =
        destroy(scope, artifactId, reason ?: "retention-expired")

    override fun setLegalHold(
        scope: TenantScope,
        artifactId: String,
        legalHold: Boolean,
        reason: String?,
    ): ArtifactMetadata? {
        val nowInstant = now()
        val nowTs = Timestamp.from(nowInstant)
        val updated = jdbcTemplate.update(
            """
            UPDATE artifacts
               SET legal_hold = ?, legal_hold_reason = ?, legal_hold_set_at = ?, updated_at = ?
             WHERE organization_id = ? AND workstream_id = ? AND artifact_id = ?
            """.trimIndent(),
            legalHold,
            if (legalHold) reason else null,
            if (legalHold) nowTs else null,
            nowTs,
            scope.organizationId,
            scope.workstreamId,
            artifactId,
        )
        if (updated == 0) return null
        val row = selectRow(scope, artifactId) ?: return null
        return ArtifactGovernance.refreshArtifactMetadata(toMetadata(row), nowInstant)
    }

    override fun collectOrphanedUploads(scope: TenantScope): List<String> {
        val keys = blobClient.listObjectKeys("$uploadPrefix/")
        val reclaimed = mutableListOf<String>()
        for (key in keys) {
            if (bestEffortDelete(key)) reclaimed.add(key)
        }
        return reclaimed
    }

    override fun listGcMetadataRows(scope: TenantScope): List<ArtifactGcMetadataRow> =
        jdbcTemplate.query(
            """
            SELECT artifact_id, storage_key, availability_status
              FROM artifacts
             WHERE organization_id = ? AND workstream_id = ?
            """.trimIndent(),
            { rs, _ ->
                ArtifactGcMetadataRow(
                    artifactId = rs.getString("artifact_id"),
                    storageKey = rs.getString("storage_key"),
                    availabilityStatus = toAvailabilityStatus(rs.getString("availability_status")),
                )
            },
            scope.organizationId,
            scope.workstreamId,
        )

    private fun destroy(scope: TenantScope, artifactId: String, reason: String): Boolean {
        val row = selectRow(scope, artifactId) ?: return false
        val nowInstant = now()
        val metadata = ArtifactGovernance.refreshArtifactMetadata(toMetadata(row), nowInstant)
        if (!ArtifactGovernance.isArtifactDestroyable(metadata, nowInstant)) return false
        val purged = purgeRow(scope, artifactId, reason, nowInstant)
        if (!purged) return false
        bestEffortDelete(row.storageKey)
        return true
    }

    private fun saveMetadata(scope: TenantScope, metadata: ArtifactMetadata, storageKey: String) {
        val nowTs = Timestamp.from(metadata.createdAt)
        val payload = objectMapper.writeValueAsString(
            linkedMapOf<String, Any?>().apply {
                put("owner", metadata.owner)
                metadata.retentionDays?.let { put("retentionDays", it) }
            },
        )
        jdbcTemplate.update(
            """
            INSERT INTO artifacts (
              organization_id, workstream_id, namespace_id, workflow_id, artifact_id,
              availability_status, retention_status, legal_hold, retention_until, purged_at,
              purge_reason, legal_hold_reason, legal_hold_set_at, content_hash, size,
              content_type, storage_key, payload, created_at, updated_at
            ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?::jsonb, ?, ?)
            ON CONFLICT (organization_id, workstream_id, namespace_id, workflow_id, artifact_id)
            DO UPDATE SET
              availability_status = EXCLUDED.availability_status,
              retention_status = EXCLUDED.retention_status,
              legal_hold = EXCLUDED.legal_hold,
              retention_until = EXCLUDED.retention_until,
              purged_at = EXCLUDED.purged_at,
              purge_reason = EXCLUDED.purge_reason,
              legal_hold_reason = EXCLUDED.legal_hold_reason,
              legal_hold_set_at = EXCLUDED.legal_hold_set_at,
              content_hash = EXCLUDED.content_hash,
              size = EXCLUDED.size,
              content_type = EXCLUDED.content_type,
              storage_key = EXCLUDED.storage_key,
              payload = EXCLUDED.payload,
              updated_at = EXCLUDED.updated_at
            """.trimIndent(),
            scope.organizationId,
            scope.workstreamId,
            scopeFor(scope, metadata.owner).namespaceId,
            scopeFor(scope, metadata.owner).workflowId,
            metadata.id,
            metadata.availabilityStatus.wireValue,
            metadata.retentionStatus.wireValue,
            metadata.legalHold,
            metadata.retentionUntil?.let { Timestamp.from(it) },
            metadata.purgedAt?.let { Timestamp.from(it) },
            metadata.purgeReason,
            metadata.legalHoldReason,
            metadata.legalHoldSetAt?.let { Timestamp.from(it) },
            metadata.hash,
            metadata.size,
            metadata.contentType,
            storageKey,
            payload,
            nowTs,
            nowTs,
        )
    }

    private fun purgeRow(scope: TenantScope, artifactId: String, reason: String, nowInstant: Instant): Boolean {
        val nowTs = Timestamp.from(nowInstant)
        val updated = jdbcTemplate.update(
            """
            UPDATE artifacts
               SET availability_status = 'purged', purged_at = ?, purge_reason = ?, updated_at = ?
             WHERE organization_id = ? AND workstream_id = ? AND artifact_id = ?
               AND legal_hold = FALSE AND availability_status <> 'purged'
            """.trimIndent(),
            nowTs,
            reason,
            nowTs,
            scope.organizationId,
            scope.workstreamId,
            artifactId,
        )
        return updated > 0
    }

    private fun selectRow(scope: TenantScope, artifactId: String): ArtifactRow? {
        val rows = jdbcTemplate.query(
            """
            SELECT organization_id, workstream_id, namespace_id, workflow_id, artifact_id,
                   availability_status, retention_status, legal_hold, retention_until, purged_at,
                   purge_reason, legal_hold_reason, legal_hold_set_at, content_hash, size,
                   content_type, storage_key, payload::text AS payload, created_at, updated_at
              FROM artifacts
             WHERE organization_id = ? AND workstream_id = ? AND artifact_id = ?
            """.trimIndent(),
            ROW_MAPPER,
            scope.organizationId,
            scope.workstreamId,
            artifactId,
        )
        return rows.firstOrNull()
    }

    private fun toMetadata(row: ArtifactRow): ArtifactMetadata {
        val payload = parsePayload(row.payloadJson)
        val owner = (payload["owner"] as? String) ?: row.namespaceId
        val retentionDays = (payload["retentionDays"] as? Number)?.toInt()
        return ArtifactMetadata(
            id = row.artifactId,
            owner = owner,
            hash = row.contentHash,
            size = row.size,
            contentType = row.contentType,
            availabilityStatus = row.availabilityStatus,
            retentionStatus = toRetentionStatus(row.retentionStatus),
            legalHold = row.legalHold,
            createdAt = row.createdAt,
            retentionDays = retentionDays,
            retentionUntil = row.retentionUntil,
            purgedAt = row.purgedAt,
            purgeReason = row.purgeReason,
            legalHoldReason = row.legalHoldReason,
            legalHoldSetAt = row.legalHoldSetAt,
        )
    }

    private fun parsePayload(json: String?): Map<String, Any?> {
        if (json.isNullOrBlank()) return emptyMap()
        return try {
            @Suppress("UNCHECKED_CAST")
            objectMapper.readValue(json, Map::class.java) as Map<String, Any?>
        } catch (_: Exception) {
            emptyMap()
        }
    }

    private fun bestEffortDelete(key: String): Boolean =
        try {
            blobClient.deleteObject(key)
        } catch (_: Exception) {
            false
        }

    private fun contentKey(hash: String): String = "$objectPrefix/${ArtifactHash.digest(hash)}"

    companion object {
        const val DEFAULT_UPLOAD_PREFIX = "uploads"
        const val DEFAULT_OBJECT_PREFIX = "objects"
        const val DEFAULT_NAMESPACE_ID = "default"
        const val DEFAULT_WORKFLOW_ID = "default"

        /**
         * Derives the namespace / workflow scope from an owner structured as
         * `namespace/workflow`. A bare owner is used as the namespace and the
         * workflow defaults to [DEFAULT_WORKFLOW_ID].
         */
        fun scopeFor(scope: TenantScope, owner: String): ArtifactRowScope = parseOwnerScope(owner, scope)

        private fun parseOwnerScope(owner: String, scope: TenantScope): ArtifactRowScope {
            if (owner.isEmpty()) {
                return ArtifactRowScope(scope.organizationId, scope.workstreamId, DEFAULT_NAMESPACE_ID, DEFAULT_WORKFLOW_ID)
            }
            val separator = owner.indexOf('/')
            return if (separator > 0 && separator < owner.length - 1) {
                ArtifactRowScope(
                    scope.organizationId,
                    scope.workstreamId,
                    owner.substring(0, separator),
                    owner.substring(separator + 1),
                )
            } else {
                ArtifactRowScope(scope.organizationId, scope.workstreamId, owner, DEFAULT_WORKFLOW_ID)
            }
        }

        private fun toAvailabilityStatus(value: String?): ArtifactAvailabilityStatus = when (value) {
            "available" -> ArtifactAvailabilityStatus.AVAILABLE
            "purged" -> ArtifactAvailabilityStatus.PURGED
            "pending" -> ArtifactAvailabilityStatus.PENDING
            "uploading" -> ArtifactAvailabilityStatus.UPLOADING
            "unavailable" -> ArtifactAvailabilityStatus.UNAVAILABLE
            else -> ArtifactAvailabilityStatus.ARCHIVED
        }

        private fun toRetentionStatus(value: String?): ArtifactRetentionStatus =
            if (value == ArtifactRetentionStatus.EXPIRED.wireValue) {
                ArtifactRetentionStatus.EXPIRED
            } else {
                ArtifactRetentionStatus.ACTIVE
            }

        private val ROW_MAPPER = RowMapper { rs: ResultSet, _: Int ->
            ArtifactRow(
                organizationId = rs.getString("organization_id"),
                workstreamId = rs.getString("workstream_id"),
                namespaceId = rs.getString("namespace_id"),
                workflowId = rs.getString("workflow_id"),
                artifactId = rs.getString("artifact_id"),
                availabilityStatus = toAvailabilityStatus(rs.getString("availability_status")),
                retentionStatus = rs.getString("retention_status"),
                legalHold = rs.getBoolean("legal_hold"),
                retentionUntil = rs.getTimestamp("retention_until")?.toInstant(),
                purgedAt = rs.getTimestamp("purged_at")?.toInstant(),
                purgeReason = rs.getString("purge_reason"),
                legalHoldReason = rs.getString("legal_hold_reason"),
                legalHoldSetAt = rs.getTimestamp("legal_hold_set_at")?.toInstant(),
                contentHash = rs.getString("content_hash"),
                size = rs.getLong("size"),
                contentType = rs.getString("content_type"),
                storageKey = rs.getString("storage_key"),
                payloadJson = rs.getString("payload"),
                createdAt = rs.getTimestamp("created_at").toInstant(),
                updatedAt = rs.getTimestamp("updated_at").toInstant(),
            )
        }
    }
}

/** Full tenant / hierarchy scope of an artifact row. */
data class ArtifactRowScope(
    val organizationId: String,
    val workstreamId: String,
    val namespaceId: String,
    val workflowId: String,
)
